src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java

```diff
@@ -366,101 +343,121 @@ protected void appendFailurePrefix(StringBuilder failureMessage, String netID, i
+            PayeeACHData achData = new PayeeACHData(payee, achDetail, achAccount.getPayeeIdentifierTypeCode(), achAccount);
+            String accountUpdateErrors = createAndRoutePayeeACHAccountDocument(achData, this::setupDocumentForACHUpdate);
+            processingResults.append(accountUpdateErrors);
         } else {
-            processingResults.append(" Input file's account information for payee of type '")
-                .append(achAccount.getPayeeIdentifierTypeCode())
-                .append("' matches what is already in KFS; no updates will be made for this entry.");
+            LOG.info("updateACHAccountIfNecessary: Input file's account information for payee of type '" + achAccount.getPayeeIdentifierTypeCode()
+                    + "' matches what is already in KFS; no updates will be made for this entry.");
+        }
+        
+        if (processingResults.length() > 0) {
+            processingResults.append(" Update was NOT performed.");
+            LOG.warn("updateACHAccountIfNecessary: " + processingResults.toString());
         }
-        processingResults.append(" Update was NOT performed.");
-        LOG.warn("updateACHAccountIfNecessary: " + processingResults.toString());
+        
         return processingResults.toString();
     }
 
+    protected boolean accountHasChanged(PayeeACHAccountExtractDetail achDetail, PayeeACHAccount achAccount) {
+        return !StringUtils.equals(achDetail.getBankRoutingNumber(), achAccount.getBankRoutingNumber())
+                || !StringUtils.equals(achDetail.getBankAccountNumber(), achAccount.getBankAccountNumber())
+                || !StringUtils.equals(
+                        getACHTransactionCode(achDetail.getBankAccountType()), achAccount.getBankAccountTypeCode());
+    }
+
+    protected void setupDocumentForACHUpdate(MaintenanceDocument paatDocument, PayeeACHData achData) {
+        if (!achData.hasOldAccount()) {
+            throw new RuntimeException("An existing ACH account should have been present");
+        }
+        PayeeACHAccountExtractDetail achDetail = achData.getAchDetail();
+        PayeeACHAccount oldAccount = achData.getOldAccount();
+        PayeeACHAccount newAccount = (PayeeACHAccount) ObjectUtils.deepCopy(oldAccount);
+        
+        newAccount.setBankRoutingNumber(achDetail.getBankRoutingNumber());
+        newAccount.setBankAccountNumber(achDetail.getBankAccountNumber());
+        newAccount.setBankAccountTypeCode(getACHTransactionCode(achDetail.getBankAccountType()));
+        newAccount.setActive(true);
+        
+        paatDocument.getOldMaintainableObject().setDataObject(oldAccount);
+        paatDocument.getNewMaintainableObject().setDataObject(newAccount);
+        paatDocument.getNewMaintainableObject().setMaintenanceAction(KFSConstants.MAINTENANCE_EDIT_ACTION);
+    }
+
+    protected String createAndRoutePayeeACHAccountDocument(
+            PayeeACHData achData, BiConsumer<MaintenanceDocument, PayeeACHData> documentConfigurer) {

```

Comment:
What is the advantage of using the BiConsumer in this case?

Reply:
I was trying to set up a more general method for creating and routing PAAT documents, and to have most of the case-specific setup deferred to a lambda expression or method reference. Using a BiConsumer allows for passing the required data to the method reference without needing to encapsulate the document within the helper PayeeACHData object.